doc/adr/0002-steppable-cek-machine.md

```diff
@@ -0,0 +1,235 @@
+
+### Abstracting our the computation to "steps"
+
+This abstraction has been implemented in [PR#4909](https://github.com/input-output-hk/plutus/pull/4909/).
+
+The current machine inlined the steps. We separate each steps into separate functions. They all return a `CekState`:
+
+```haskell
+data CekState uni fun =
+    -- the next state is computing
+    Computing WordArray (Context uni fun) (Closure uni fun)
+    -- the next state is returning
+    | Returning WordArray (Context uni fun) (CekValue uni fun)
+    -- evaluation finished
+    | Terminating (Term NamedDeBruijn uni fun ())
+
+data Closure uni fun = 
+  Closure (Term NamedDeBruijn uni fun ()) (CekValEnv uni fun)
+```
+
+The computing step is `computeCekStep` with the following signature:
+
+```haskell
+computeCekStep
+    :: forall uni fun s
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    => WordArray
+    -> Context uni fun
+    -> Closure uni fun
+    -> CekM uni fun s (CekState uni fun)
+```
+
+Similarly for the returning step (`returnCekStep`). Then we link up all the steps with `continue`, and the machine behaves very similar to our current one:
+
+```haskell
+continue :: forall uni fun s
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    => CekState uni fun
+    -> CekM uni fun s (Term NamedDeBruijn uni fun ())
+continue (Computing !unbudgetedSteps ctx (Closure term env)) = do
+    state <- computeCekStep unbudgetedSteps ctx env term
+    continue state
+continue (Returning !unbudgetedSteps ctx val) = do
+    state <- returnCekStep unbudgetedSteps ctx val
+    continue state
+continue (Terminating term) = pure term
+```
+
+### Alternative: use CPS
+
+This approach has been suggested to improve the performance of the debugging machine. The abstraction proposed above does not have good performance. To improve performance, we need to get rid of the creation of the state objects. This is because GHC probably doesn't have the capability to eliminate those, because the place where they're destructed is too far from where they're constructed. That means that instead we probably need to do something with continuations:
+
+```haskell
+{-# INLINE computeStep #-}
+computeStep :: (... -> r) -> .... -> r
+computeStep k =
+  go ...
+    -- instead of Computing $ ...
+    k ...
+```
+
+So `computeStep` gets inlined, as before, at which point we hopefully know `k`, so GHC can optimize it properly. Except it's going to be more complicated than that because:
+
+1. We need to pass continuations for all the different cases, so three of them.
+2. The continuations are recursive with the main function! It's not clear how to make it work nicely.
+
+We currently aren't taking this approach because:
+
+- The performance of the debugging machine is not a main concern.
+- It's unclear how to make some parts work nicely.
+
+If we find out that the debugging machine is too slow, we may want to revisit this approach again.
+
+### Coroutines in Haskell
```

Comment:
Do we have a prototype of this? I'd rather see a prototype before deciding on this based on an abstract argument.

Reply:
I also don't think this needs to go in here: if the decision is to have a separate machine, we can have a separate ADR once we decide how to structure it.

Reply:
OK it sounds like this (implementing a separate machine) will be the alternative anyway, so we're not deciding on that now. I'd rather include this as the alternative since it's already written anyway. We can of course have another ADR when it comes to the point that we have to revisit the alternative.

Reply:
Sorry, I didn't quite follow that... What I'm suggesting is that we have one ADR for whether or not to use a separate machine, and one for how to structure the steppable machine, however we do it. Note that what you've written here about coroutines will I think be useful regardless, but it definitely belongs to a discussion of how to structure the steppable machine.

Reply:
OK I see. I can do that.

Reply:
I'll open another PR for another ADR - this one will stay talking about how to structure the steppable machine. The new PR will have the ADR about how to have the debugging machine share code with the production machine without compromising performance.

Reply:
Yeah I think that will be a lot clearer. Great suggestion!